### PR TITLE
Diagnose misuse of the cleanup attribute

### LINKED ISSUE #79443: Diagnose misuse of the cleanup attribute
The cleanup attribute is written on a variable declaration, and the address of the declared variable is passed to the specified function in the attribute. This has some subtle issues we should diagnose on. Consider:
```
__attribute__((cleanup(free))) void *p = malloc(10);
```
This will pass `&p` to `free`, which is going to do the wrong thing. GCC diagnoses this kind of misuse (`warning: 'free' called on unallocated object 'p' [-Wfree-nonheap-object]`) and it would be good if we did the same: https://godbolt.org/z/nhqh9EazM

### PR DESCRIPTION
**Pull request by 11happy**
**Overview:**
This pull request fixes #79443 when the cleanup attribute is intended to be applied to a variable declaration, passing its address to a specified function. The problem arises when standard functions like free, closedir, fclose, etc., are used incorrectly with this attribute, leading to incorrect behavior. 

**Testing:**
- Tested the updated code.
- Verified that other functionalities remain unaffected.

Screenshot from 2024-01-30 23-15-51

Screenshot from 2024-01-30 23-22-17

**Dependencies:**
- No dependencies on other pull requests.

**CC:**
- @AaronBallman

### CODE REVIEW

**Comment by 11happy** [comment]
the current working is as expected  as per my understanding, However I request comments & suggestions from others as I am not quite sure that this would be fine,also apologies for not so good naming.
Thank you

**Comment by 11happy** [comment]
I was going through the thread on discord whether we should use same diagnostics as GCC or not, so as suggested there we should not directly copy but having same essence will be fine. so in recent commit changed the warning.
Thank you

**Review by erichkeane — COMMENTED** [review]
This needs a release note and some tests as well.

**Review by AaronBallman — COMMENTED** [review]
Thank you for working on this! I think we want to go in a different direction for this fix. I think the real issue is that we're missing a call to `Sema::CheckFunctionCall()` within `handleCleanupAttr()`. That's currently a private function in `Sema`, so it would have to be hoisted to a public interface so we can call it. But after we call `Sema::CheckAssignmentConstraints()`, I think we should make a fake function call expression so that we can call `Sema::CheckFunctionCall()` on it.

That code should look something like this (note, probably needs more work to ensure source locations point to the correct things for correct diagnostic behavior):
```
  DeclRefExpr *FakeVarRef = DeclRefExpr::Create(
      S.Context, NestedNameSpecifierLoc{}, SourceLocation{}, VD, false,
      DeclarationNameInfo{VD->getDeclName(), VD->getLocation()}, VD->getType(),
      VK_LValue);
  Expr *FakeArg = UnaryOperator::Create(
      S.Context, FakeVarRef, UnaryOperatorKind::UO_AddrOf,
      S.Context.getPointerType(VarDecl->getType()), VK_PRValue, OK_Ordinary,
      SourceLocation{}, false, FPOptionsOverride{});
  CallExpr *FakeCall =
      CallExpr::Create(S.Context, E, ArrayRef{FakeArg}, S.Context.VoidTy,
                       VK_PRValue, SourceLocation{}, FPOptionsOverride{});
  S.CheckFunctionCall(FD, FakeCall, FD->getType()->getAs<FunctionProtoType>());
```
Then we get all of the checking functionality that the user would get had they called the function themselves directly. This includes warning about things like this:
```
struct __attribute__((aligned(8))) S {};

void derp(struct S *ptr);

int main(void) {
  typedef __attribute__((aligned(4))) struct S TypedefAligned4;
  TypedefAligned4 s1;
  derp(&s1); // Warns

  __attribute__((cleanup(derp))) TypedefAligned4 s2; // Should get same warning here as above
}
```

The changes should also come with a release note in `clang/docs/ReleaseNotes.rst` so users know about the fix, and should come with tests in `clang/test/Sema/` to demonstrate the behavioral changes.

**Comment by 11happy** [comment]
tried to implement the requested changes along the new direction, further I have some queries should I let this statament as is or assign to a variable:
```
S.CheckFunctionCall(FD, FunctionCallExpression,
                      FD->getType()->getAs<FunctionProtoType>());
```
Further I request some guidance on how to use `malloc` & `free` etc , standard library functions in tests.
Added Release notes but not sure about its framing.

**Comment by AaronBallman** [comment]
> tried to implement the requested changes along the new direction, I have some queries should I let this statament as is or assign to a variable:
> 
> ```
> S.CheckFunctionCall(FD, FunctionCallExpression,
>                       FD->getType()->getAs<FunctionProtoType>());
> ```

That call only returns `false` but I would still wrap the call in an `if` statement, as in:
```
if (S.CheckFunctionCall(...))
  return;
```
and I would move the code up above where we call `addAttr()`. The basic idea is: if `CheckFunctionCall()` ever gets updated to return `true` on failure, then we'll skip adding the problematic attribute to the declaration.

> Further I request some guidance on how to use `malloc` & `free` etc , standard library functions in tests. Added Release notes but not sure about its framing.

We want our tests to be hermetic (not relying on anything specific to the developer's setup), so tests for this sort of thing will declare `free` manually. e.g,
```
extern void free(void *);
```

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -106,6 +106,9 @@ Improvements to Clang's diagnostics
 - Clang now applies syntax highlighting to the code snippets it
   prints.
 
+- Clang now provides improved warnings for the cleanup attribute to detect misuse scenarios,
+  such as attempting to call `free` on unallocated objects.
```
Comment (AaronBallman) [inline clang/docs/ReleaseNotes.rst:110]:
```suggestion
- Clang now provides improved warnings for the ``cleanup`` attribute to detect misuse scenarios,
  such as attempting to call ``free`` on an unallocated object. Fixes
  `#79443 <https://github.com/llvm/llvm-project/issues/79443>`_.
```

**Review by AaronBallman — COMMENTED** [review]
I think the only thing left is to add test coverage somewhere in `clang/test/Sema/`, probably in https://github.com/llvm/llvm-project/blob/main/clang/test/Sema/attr-cleanup.c.

You can probably test with something like:
```
extern void free(void *);
extern void *malloc(__typeof__(sizeof(0)));

void foo(void) {
  __attribute__((cleanup(free))) void *p = malloc(10);
}
```
but additional test coverage would make sense such as what I pointed out here: https://github.com/llvm/llvm-project/pull/80040#pullrequestreview-1856764787

**Comment by 11happy** [comment]
> extern void free(void *);

Sure I already wrote some tests and will commit after locally checking them  but idk everytime I restart pc clang rebuilds itself and that is taking time, is it normal for clang to rebuild everytime I restart? (I think it has something to do with the dual boot setup linux/windows.)

**Comment by 11happy** [comment]
```
extern void free(void *);
extern void *malloc(size_t size);
void t8(void) {
  void *p __attribute__((cleanup(free))) = malloc(10); // expected-warning{{attempt to call free on non-heap object 'p'}}
}
```
 I added  this, but 
I am getting this error: (Sorry if its a silly doubt)

```
error: 'expected-warning' diagnostics expected but not seen: 
  File /home/happy/LLVM/llvm-project/clang/test/Sema/attr-cleanup.c Line 55: attempt to call free on non-heap object 'p'
error: 'expected-warning' diagnostics seen but not expected: 
  (frontend): attempt to call free on non-heap object 'p'
2 errors generated.
```

**Comment by erichkeane** [comment]
> ```
> extern void free(void *);
> extern void *malloc(size_t size);
> void t8(void) {
>   void *p __attribute__((cleanup(free))) = malloc(10); // expected-warning{{attempt to call free on non-heap object 'p'}}
> }
> ```
> 
> I added this, but I am getting this error: (Sorry if its a silly doubt)
> 
> ```
> error: 'expected-warning' diagnostics expected but not seen: 
>   File /home/happy/LLVM/llvm-project/clang/test/Sema/attr-cleanup.c Line 55: attempt to call free on non-heap object 'p'
> error: 'expected-warning' diagnostics seen but not expected: 
>   (frontend): attempt to call free on non-heap object 'p'
> 2 errors generated.
> ```

That looks like you got the source-location wrong on your diagnostic.

**Comment by 11happy** [comment]
Umm how could I resolve that? I tried various things but none working. 🥲

**Comment by erichkeane** [comment]
> Umm how could I resolve that? I tried various things but none working. 🥲

You'll have to spend some time in a debugger (note that `clang::DiagnosticIDs::EmitDiag` is where the diagnostic happens, so is a good place for a brekapoint).  But I see that in your creation of your expressions you use `SourceLocation{}` in a few places (which is the 'empty' source `SourceLocation` and likely what is being used for the diagnostic).

So you'll likely need to come up with a good location for at least one of those.

**Comment by 11happy** [comment]
Got it

**Comment by 11happy** [comment]
I have added a test, but will be adding more in some upcoming commits as I test them locally.

**Comment by 11happy** [comment]
I have fixed the merge conflict , will be adding more tests as I am able to test them locally.
Thank you

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 2]
```diff
@@ -3780,6 +3780,30 @@ static void handleCleanupAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+      S.Context, NestedNameSpecifierLoc{}, SourceLocation{Loc}, VD, false,
+      DeclarationNameInfo{VD->getDeclName(), VD->getLocation()}, VD->getType(),
+      VK_LValue);
+
+  // Create a unary operator expression that represents taking the address of
+  // the variable. This is a fake/dummy expression.
+  Expr *AddressOfVariable = UnaryOperator::Create(
+      S.Context, VariableReference, UnaryOperatorKind::UO_AddrOf,
+      S.Context.getPointerType(VD->getType()), VK_PRValue, OK_Ordinary,
+      SourceLocation{Loc}, false, FPOptionsOverride{});
```
Comment (erichkeane) [inline clang/lib/Sema/SemaDeclAttr.cpp:3796]:
```suggestion
      Loc, false, FPOptionsOverride{});
```

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 3]
```diff
@@ -3780,6 +3780,30 @@ static void handleCleanupAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
       << NI.getName() << ParamTy << Ty;
     return;
   }
+  VarDecl *VD = cast<VarDecl>(D);
+  // Create a reference to the variable declaration. This is a fake/dummy
+  // reference.
+  DeclRefExpr *VariableReference = DeclRefExpr::Create(
+      S.Context, NestedNameSpecifierLoc{}, SourceLocation{Loc}, VD, false,
```
Comment (erichkeane) [inline clang/lib/Sema/SemaDeclAttr.cpp:3787]:
```suggestion
      S.Context, NestedNameSpecifierLoc{}, Loc, VD, false,
```

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 4]
```diff
@@ -3780,6 +3780,30 @@ static void handleCleanupAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+  // the variable. This is a fake/dummy expression.
+  Expr *AddressOfVariable = UnaryOperator::Create(
+      S.Context, VariableReference, UnaryOperatorKind::UO_AddrOf,
+      S.Context.getPointerType(VD->getType()), VK_PRValue, OK_Ordinary,
+      SourceLocation{Loc}, false, FPOptionsOverride{});
+
+  // Create a function call expression. This is a fake/dummy call expression.
+  CallExpr *FunctionCallExpression = CallExpr::Create(
+      S.Context, E, ArrayRef{AddressOfVariable}, S.Context.VoidTy, VK_PRValue,
+      SourceLocation{Loc}, FPOptionsOverride{});
```
Comment (erichkeane) [inline clang/lib/Sema/SemaDeclAttr.cpp:3801]:
```suggestion
      Loc, FPOptionsOverride{});
```

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 3 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaDeclAttr.cpp:3787]: Also note, Loc is the location of the attribute itself, and is probably not what you want to make the location here.  I would THINK we want the diagnostic to point to the function instead... WDYT?

Comment (11happy) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Yes I agree with you , should we use `FD->getlocation()` instead?

Comment (erichkeane) [reply clang/lib/Sema/SemaDeclAttr.cpp]: I think the function decl makes more sense for the diag... What do you think @AaronBallman ?

Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Making sure we're all on the same page, I think the diagnostic should show up like:
```
extern void free(void *);
extern void *malloc(size_t size);
void t8(void) {
  void *p
  __attribute__((
	  cleanup(
		  free // expected-warning{{attempt to call free on non-heap object 'p'}}
	  )
	))
  = malloc(10);
```
and
```
struct __attribute__((aligned(8))) S {};

void derp(struct S *ptr);

int main(void) {
  typedef __attribute__((aligned(4))) struct S TypedefAligned4;
  TypedefAligned4 s1;
  derp(&s1); // expected-warning {{passing 4-byte aligned argument to 8-byte aligned parameter 1 of 'derp' may result in an unaligned pointer access}}

  __attribute__((cleanup(derp))) // expected-warning {{passing 4-byte aligned argument to 8-byte aligned parameter 1 of 'derp' may result in an unaligned pointer access}}
  TypedefAligned4 s2;
}
```

Comment (11happy) [reply clang/lib/Sema/SemaDeclAttr.cpp]: done

**Comment by 11happy** [comment]
I have added the tests , and its working as expected . Clang-diagnostics is indeed interesting to work on.
Thank you

**Comment by 11happy** [comment]
sorry for last 3 continuous format code commits, I usually run `git clang-format HEAD~1` and also ran this time idk why it was not correctly formatted so I tried manually and it resulted in so many commits,
Thank you

File: clang/test/Sema/attr-cleanup.c [hunk 5]
```diff
@@ -1,7 +1,7 @@
-// RUN: %clang_cc1 %s -verify -fsyntax-only
+// RUN: %clang_cc1 -Wfree-nonheap-object -fsyntax-only -verify %s
 
 void c1(int *a);
-
+typedef unsigned long size_t;
```
Comment (AaronBallman) [inline clang/test/Sema/attr-cleanup.c:4]:
```suggestion
typedef __typeof__(sizeof(0)) size_t;
```

**Review by AaronBallman — COMMENTED** [review]
Precommit CI found a valid failure on Windows: 
```
FAIL: Clang :: Sema/attr-cleanup.c (15237 of 19292)
******************** TEST 'Clang :: Sema/attr-cleanup.c' FAILED ********************
Exit Code: 1
Command Output (stdout):
--
# RUN: at line 1
c:\ws\src\build\build-clang-windows\bin\clang.exe -cc1 -internal-isystem C:\ws\src\build\build-clang-windows\lib\clang\19\include -nostdsysteminc -Wfree-nonheap-object -fsyntax-only -verify C:\ws\src\clang\test\Sema\attr-cleanup.c
# executed command: 'c:\ws\src\build\build-clang-windows\bin\clang.exe' -cc1 -internal-isystem 'C:\ws\src\build\build-clang-windows\lib\clang\19\include' -nostdsysteminc -Wfree-nonheap-object -fsyntax-only -verify 'C:\ws\src\clang\test\Sema\attr-cleanup.c'
# .---command stderr------------
# | error: 'expected-warning' diagnostics seen but not expected:
# |   File C:\ws\src\clang\test\Sema\attr-cleanup.c Line 53: incompatible redeclaration of library function 'malloc'
# | error: 'expected-note' diagnostics seen but not expected:
# |   File C:\ws\src\clang\test\Sema\attr-cleanup.c Line 53: 'malloc' is a builtin with type 'void *(unsigned long long)'
# | 2 errors generated.
# `-----------------------------
# error: command failed with exit status: 1
```
I left a suggestion on how to correct this.

**Review by erichkeane — APPROVED** [review]

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Comment by 11happy** [comment]
Sorry, I left the previous function declaration in the `sema.h`, I have removed it now.

**Comment by 11happy** [comment]
what would be the reason for windows build failing , is it a CI issue or specific to this PR & what can I do to resolve that.
Thank you

**Comment by erichkeane** [comment]
> what would be the reason for windows build failing , is it a CI issue or specific to this PR & what can I do to resolve that. Thank you

That appears to be a problem with the CI itself.  I think we've fixed up a bunch of the CI, but it will require doing a 'merge' with main.  That said, I think you're OK to just squash/merge as-is.

**Review by shafik — COMMENTED** [review]
The failure of the Windows CI is due to a know problem w/ false detection of a virus. You can use `--allow-empty` to make an empty commit and restart the process. You have approval, so once it goes green you should be good to squash and merge.

**Comment by shafik** [comment]
It looks like it passed on your last commit but you have a conflict now which you need to resolve.

Can you merge or do you need help with that?

**Review by Endilll — COMMENTED** [review]
`Sema.h` changes look good to me.

**Pull request merged**